// osuapi-js/classes/beatmap.js
const utils = require('../utils/beatmaps');
const { get } = require('../utils/request');
const { parseModeID } = require('../utils/random');
const { parseContent } = require('osu-parser');
const PPcalculator = require('osu-pp-calculator');
const { calcMods } = require('../utils/mods');

class Beatmap {
    constructor(info) {
        /**
         * @private
         */
        this.api = info.api;

        this.diff    = { };
        this.ID      = Number(info.beatmap_id);
        this.mapsetID= Number(info.beatmapset_id);
        this.mode    = parseModeID(info.mode);
        this.modeID  = Number(info.mode);
        
        utils.mergeInfo(info, this);
        utils.mergeDiffInfo(info, this.diff);
    }

    async getMapset() {
        return await this.api.getBeatmap(this.mapset, null, true);
    }

    async getScores(limit = 50, mode = this.mode, mods, user = null, type = null) {
        if(this.diff.statusID < 1)
            throw new ReferenceError('Can\'t get scores of a map that isn\'t ranked!');
        
        if(!mode)
            mode = this.mode;
        
        return (await this.api.getScores(this.ID, limit, mode, mods, user, type))
            .map(score => { score.beatmap = this; return score });
    }

    calcPP(accuracy = 100, mods = [], combo = this.diff.combo, misses = 0) {
        if(!this.api.options.baseURI.includes('osu.ppy.sh'))
            throw new Error('Can\'t calculate pp for servers other than Bancho');

        accuracy = accuracy[accuracy.length - 1] === '%' ? accuracy.substr(0, accuracy.length - 1) : accuracy;
        if(typeof accuracy !== 'number' && isNaN(accuracy))
            throw new TypeError('accuracy has to be a number');
        if(accuracy > 100 || accuracy < 0)
            throw new RangeError('accuracy has to be in range between 0 and 100');

        if(!mods)
            mods = [];
        if(mods && !Array.isArray(mods))
            mods = [ mods ];

        if(!combo)
            combo = this.diff.combo;
        if(typeof combo !== 'number' && isNaN(combo))
            throw new TypeError('combo has to be a number');
        if(combo < 1 || combo > this.diff.combo)
            throw new RangeError('combo has to be in range between 1 and beatmap\'s max combo');
            
        if(typeof misses !== 'number' && isNaN(misses))
            throw new TypeError('misses has to be a number');

        if(misses > 0 && (!combo || combo == this.diff.combo))
            throw new TypeError('You can\'t miss and still get an fc lol');

        return new Promise((resolve, reject) => {
            get('https://osu.ppy.sh/osu/' + this.ID, (err, res, body) => {
                if(err) return reject(err);
                body = parseContent(body);
                body = PPcalculator.Beatmap.fromOsuParserObject(body);
                body.mode = this.modeID;
                const pp = PPcalculator
                    .PPCalculator.calculate(body, accuracy, calcMods(mods), combo, misses);

                resolve(pp);
            });
        });
    }
}

module.exports = Beatmap;
