// osuapi-js/classes/events.js
class EpicEvent {
    constructor(info) {
        this.html       = info.display_html; //(new DOMParser()).parseFromString(info.display_html, 'text/html');
        this.date       = new Date(info.date + ' UTC+8');
        this.epicFactor = Number(info.epicfactor);
    }
}

class RankAchieved extends EpicEvent {
    constructor(info) {
        super(info);

        const html  = info.display_html;
        this.mode   = info.mode;

        const rank1 = html.indexOf('#') + 1;
        const rank2 = html.indexOf('<', rank1);
        const rank3 = html.indexOf(' ', rank1);
        this.rank   = html.slice(rank1, rank2 < rank3 ? rank2 : rank3);

        this.rating = html.slice(html.indexOf('/', 16) + 1, html.indexOf('_', 16));

        this.beatmap= info.beatmap;
    }

    static async newRA(info, api) {
        let mode = info.display_html.slice(
            info.display_html.lastIndexOf('(') + 1, info.display_html.lastIndexOf(')'));
        switch(mode) {
            case 'osu!':
                mode = 'standard';
                break;
            case 'Taiko':
                mode = 'taiko';
                break;
            case 'osu!mania':
                mode = 'mania';
                break;
            case 'Catch the Beat':
                mode = 'ctb';
                break;
        }
        info.mode = mode;
        info.beatmap = await api.getBeatmaps(info.beatmap_id);
        return new RankAchieved(info);
    }
}

class MapsetUpdate extends EpicEvent {
    constructor(info) {
        super(info);

        this.beatmapset = info.mapset;
    }

    static async newMU(info, api) {
        info.mapset = await api.getBeatmaps(info.beatmapset_id, true, null, false);
        return new MapsetUpdate(info);
    }
}

class MedalUnlocked extends EpicEvent {
    constructor(info) {
        super(info);

        const html = info.display_html;
        this.medal = html.slice(html.lastIndexOf('<b>') + 1, html.lastIndexOf('</b>'));
    }
}

module.exports = async (e, api) => {
    return new Promise(async(resolve, reject) => {
        if(e.display_html.includes('achieved'))
            resolve(await RankAchieved.newRA(e, api));
        else if(e.display_html.includes('updated'))
            resolve(await MapsetUpdate.newMU(e, api));
        else if(e.display_html.includes('unlocked'))
            resolve(new MedalUnlocked(e));
        else
            resolve(new EpicEvent(e));
    })
}
