// osuapi-js/classes/scores.js
const { parseMods } = require('../utils/mods');

function calcAcc(count) {
    const g = count['300'];
    const k = count['100'];
    const b = count['50'];
    const m = count.miss;
    return (300 * g + 100 * k + 50 * b) / (300 * (g + k + b + m));
}

class Score {
    constructor(info) {
        this.api    = info.api;
        this.score  = Number(info.score);
        this.count  = {
            '300': Number(info.count300),
            '100': Number(info.count100),
            '50':  Number(info.count50),
            miss:  Number(info.countmiss),
            '300g':Number(info.countgeki),
            '100k':Number(info.countkatu)
        }
        this.acc    = calcAcc(this.count) * 100;
        this.rank   = info.rank;
        this.combo  = Number(info.maxcombo);
        this.date   = new Date(info.date + ' UTC+8');
        this.beatmap= info.beatmap ? info.beatmap : info.beatmap_id;
        this.user   = info.user ? info.user : { 
            username: info.username,
            id: Number(info.user_id)
        }
        this.FC     = info.perfect > 0 ? true : false;
        this.mods   = parseMods(info.enabled_mods);
    }

    async getBeatmap() {
        if(this.beatmap.constructor.name === 'Beatmap')
            return this.beatmap
        else
            return await this.api.getBeatmaps(this.beatmap, false, null, false);
    }
}

class BeatmapScore extends Score {
    constructor(info) {
        super(info);
        this.id = Number(info.score_id);
        this.pp = Number(info.pp);
    }
}

class UserScore extends Score {
    constructor(info) {
        super(info);
        this.pp = Number(info.pp);
    }
}

class MultiScore extends Score {
    constructor(info) {
        super(info);

        this.team = info.team === '0' ? null : info.team === '1' ? 'blue' : 'red';
        this.passed = info.pass > 0 ? true : false;
    }
}

module.exports = {
    BeatmapScore,
    MultiScore,
    UserScore
}
