// osuapi-js/classes/user.js
const parseEvent = require('./events');

class User {
    constructor(info) {
        this.api      = info.api;
        this.ID       = Number(info.user_id);
        this.username = info.username;
        this.count    = {
            '300':       Number(info.count300),
            '100':       Number(info.count100),
            '50':        Number(info.count50),
            play:        Number(info.playcount),
            SS:          Number(info.count_rank_ss),
            S:           Number(info.count_rank_s),
            A:           Number(info.count_rank_a)
        }
        this.score    = {
            total:       Number(info.total_score),
            ranked:      Number(info.ranked_score)
        }
        this.pp       = {
            rank:        Number(info.pp_rank),
            raw:         Number(info.pp_raw),
            countryRank: Number(info.pp_country_rank)
        }
        this.level     = Number(info.level);
        this.accuracy  = Number(info.accuracy);
        this.country   = info.country;
        this.events    = info.events;
        this.mode      = info.mode;
    }

    async getBest(limit = 10) {
        const scores = await this.api.getUserBest(this.ID, this.mode, limit, 'id');
        scores.forEach(score => score.user = this);
        return scores;
    }

    static async newUser(info) {
        for(let e in info.events) {
            let ev = info.events[e];
            info.events[e] = await parseEvent(ev, info.api);
        }
        return new User(info);
    }
}

module.exports = User;
