// osuapi-js/index.js
const { get }       = require('./utils/request');
const osuAPIOptions = require('./constants/defaultOptions');
const User          = require('./classes/user');
const Beatmap       = require('./classes/beatmap');
const Beatmapset    = require('./classes/beatmapset');
const utils         = require('./utils/random');
const Scores        = require('./classes/scores');
const { calcMods }  = require('./utils/mods');

class osuAPI {
    /**
     * @param {string} APIKey 
     * @param {osuAPIOptions} [options] 
     */
    constructor(APIKey, options = {}) {
        if(!APIKey)
            throw new TypeError('You have to provide an API key');
        if(typeof APIKey !== 'string')
            throw new TypeError('APIKey must be a string');
        /**
         * Your osu! API key
         * @type {string}
         * @private
         */
        this.key = APIKey;

        /**
         * API options
         * @type {osuAPIOptions}
         * @private
         */
        this.options = this._validateOptions(options);
        /**
         * @type {Date}
         * @private
         */
        this.lastCall = null;
    }

    /**
     * Fetch information about user.
     * @param {string | number} user Username or user ID.
     * @param {'standard' | 'taiko' | 'ctb' | 'mania'} [mode='standard']
     * Which mode gather information from? 
     * @param {number} [eventDays=1]
     * Max number of days between now and last event date. 
     * Range of 1-31. Optional, default value is 1.
     * @param {'username' | 'id'} [type='username']
     * Specify if user is a user ID or a username. 
     * Optional, default behaviour is automatic recognition 
     * (may be problematic for usernames made up of digits only).
     */
    async getUser(user, mode = 'standard', eventDays = 1, type = 'username') {
        const qs = { u: user, event_days: eventDays, m: utils.parseMode(mode), type: 'string' };
        
        if(typeof user === 'string' || type === 'username') {
            qs.type = 'string';
            user = user.replace(/ /g, '_');
        } else if((typeof user === 'number' || !isNaN(user)) || type === 'id')
            qs.type = 'id';
        else
            throw new TypeError('user has to be a string or a number');

        if(typeof eventDays !== 'number' && isNaN(eventDays))
            throw new TypeError('eventDays has to be a number');
        if(eventDays < 1 || eventDays > 31)
            throw new RangeError('eventDays has to be in range between 1 and 31');

        if(type !== 'username' && type !== 'id')
            throw new TypeError('type has to either be "username" or "id"');

        try {
            const info = (await this._apiCall('get_user', qs))[0];
            info.mode = mode;
            return await User.newUser(info);
        } catch(e) {
        //    console.log(e);
            return null;
        }
    }

    /**
     * 
     * @param {string | number} user 
     * @param {'standard' | 'taiko' | 'ctb' | 'mania'} mode 
     * @param {number} limit 
     * @param {'username' | 'id'} type 
     */
    async getUserBest(user, mode, limit = 10, type) {
        const qs = { u: user, m: utils.parseMode(mode), limit, type: 'string' };
        
        if(typeof user === 'string' || type === 'username') {
            qs.type = 'string';
            user = user.replace(/ /g, '_');
        } else if((typeof user === 'number' || !isNaN(user)) || type === 'id')
            qs.type = 'id';
        else
            throw new TypeError('user has to be a string or a number');

        if(typeof limit !== 'number' && isNaN(limit))
            throw new TypeError('limit has to be a number');
        if(limit < 1 || limit > 100)
            throw new RangeError('limit has to be in range between 1 and 100');

        const body = await this._apiCall('get_user_best', qs);
        return body.map(score => new Scores.UserScore(score));
    }

    /**
     * 
     * @param {number} ID 
     * @param {boolean} [mapset=false]
     * @param {'standard' | 'taiko' | 'ctb' | 'mania'} [mode=null]
     * @param {boolean} [fetchCreatorInfo] 
     * @returns {Promise<Beatmap>} 
     */
    async getBeatmaps(ID, mapset = false, mode = null, fetchCreatorInfo = this.options.beatmaps.fetchCreatorInfo) {
        const qs = { };

        if(typeof ID !== 'number' && isNaN(ID))
            throw new TypeError('beatmap id has to be a number');

        if(mode)
            qs.mode = utils.parseMode(mode);
            
        if(typeof mapset !== 'boolean')
            throw new TypeError('mapset has to be a boolean');

        if(mapset)
            qs.s = ID;
        else
            qs.b = ID;

        const body = await this._apiCall('get_beatmaps', qs);
        
        if(fetchCreatorInfo)
            body[0].creator = await this.getUser(body[0].creator);

        if(mapset)
            return new Beatmapset(body);
        else
            return new Beatmap(body[0]);
    }

    async getScores(beatmap, limit = 50, mode = 'standard', mods, user, type) {
        const qs = { };

        if(typeof beatmap !== 'number' && isNaN(beatmap))
            throw new TypeError('beatmap id has to be a number')
        if(typeof limit !== 'number' && isNaN(limit))
            throw new TypeError('limit has to be a number');
        qs.b = beatmap;

        if(limit < 1 || limit > 100)
            throw new RangeError('limit has to be in range between 1 and 100');
        qs.limit = limit;

        if(user) {
            if(typeof user === 'string' || type === 'username') {
                qs.type = 'string';
                user = user.replace(/ /g, '_');
            } else if((typeof user === 'number' || !isNaN(user)) || type === 'id')
                qs.type = 'id';
            else
                throw new TypeError('user has to be a string or a number');

            if(type !== 'username' && type !== 'id')
                throw new TypeError('type has to either be "username" or "id"');
            qs.u = user;
            qs.type = type;
        }

        if(mods) {
            qs.mods = calcMods(mods);
        }
        
        qs.m = utils.parseMode(mode);

        const body = await this._apiCall('get_scores', qs);

        return body.map(score => new Scores.BeatmapScore(score));
    }

    /**
     * Merges user's options with default ones and validates them
     * @param {osuAPIOptions} options 
     * @private
     */
    _validateOptions(options) {
        if(options.baseURI && (options.baseURI !== 'string' || !/http(s):\/\//.test(options.baseURI)))
            throw new TypeError('baseURI has to be an URL string');
        options.baseURI = options.baseURI || osuAPIOptions.baseURI;
        if(options.baseURI[options.baseURI.length-1] !== '/')
            options.baseURI += '/';

        if(options.beatmaps) {
            if(typeof options.beatmaps !== 'object')
                throw new TypeError('beatmapsOptions must be an object');
            for(let option in osuAPIOptions.beatmaps) {
                if(options.beatmaps[option]) {
                    if(typeof options.beatmaps[option] !== typeof osuAPIOptions.beatmaps[option])
                        throw new TypeError(`beatmaps.${option} must be a ${typeof osuAPIOptions.beatmaps[option]}`);
                } else {
                    options.beatmaps[option] = osuAPIOptions.beatmaps[option];
                }
            }
        } else
            options.beatmaps = osuAPIOptions.beatmaps;

        if(options.requestsPerMinute) {
            if(typeof options.requestsPerMinute !== 'number' && isNaN(options.requestsPerMinute))
                throw new TypeError('requestsPerMinute must be a number');
            options.requestsPerMinute = Number(options.requestsPerMinute);
        } else
            options.requestsPerMinute = osuAPIOptions.requestsPerMinute;

        return options;
    }

    /**
     * Makes an API call
     * @param {string} path
     * @param {object} qs
     * @private
     */
    async _apiCall(path, qs) {
        if(this.lastCall) {
            const delay = 1 / ((this.options.requestsPerMinute / 60) / 1000);
            const diff = new Date - this.lastCall;
            if(diff < delay)
                await utils.sleep(delay - diff);
        }
        qs.k = this.key;
        return new Promise((resolve, reject) => {
            this.lastCall = new Date;
            get(this.options.baseURI + path, (err, res, body) => {
                if(err) throw err;

                try {
                    body = JSON.parse(body);
                } catch(e) {
                    reject(e);
                }
                if(!Array.isArray(body)) {
                    if(body.error)
                        throw new Error(body.error)
                    else
                        reject(body);
                }

                if(body.length === 0)
                    resolve(null);
                else
                    body.forEach(cell => cell.api = this);

                resolve(body);
            }, { qs });
        });
    }
}

module.exports = osuAPI;
