// osuapi-js/utils/beatmaps.js
const genres = require('../constants/genres');

class BeatmapsUtils {
    static mergeInfo(info, diff) {
        diff.title   = info.title;
        diff.artist  = info.artist;
        diff.creator = info.creator;
        diff.source  = info.source ? info.source : null;
        diff.BPM     = Number(info.bpm);
        diff.tags    = info.tags.split(' ');
        diff.date    = {
            ranked:     new Date(info.approved_date + ' UTC+8'),
            lastUpdate: new Date(info.last_update + ' UTC+8')
        }
        diff.statusID= Number(info.approved);
        diff.status  = BeatmapsUtils.parseStatus(diff.statusID);
        diff.genreID = Number(info.genre_id);
        diff.genre   = genres[diff.genreID];
    }

    static mergeDiffInfo(info, diff) {
        diff.stars  = Number(info.difficultyrating);
        diff.cs     = Number(info.diff_size);
        diff.od     = Number(info.diff_overall);
        diff.ar     = Number(info.diff_approach);
        diff.hp     = Number(info.diff_drain);
        diff.name   = String(info.version);
        diff.count  = {
            play:       Number(info.playcount),
            pass:       Number(info.passcount),
            favourite:  Number(info.favourite_count)
        }
        diff.length = {
            total:      Number(info.total_length),
            hit:        Number(info.hit_length)
        }
        diff.combo  = Number(info.max_combo);
    }
    
    /**
     * Convert status id to string
     * @param {number} status 
     */
    static parseStatus(status) {
        switch(status) {
            case -2:
                return 'Graveyard';
                break;
            case -1:
                return 'WIP';
                break;
            case 0:
                return 'Pending';
                break;
            case 1:
                return 'Ranked';
                break;
            case 2:
                return 'Approved';
                break;
            case 3:
                return 'Qualified';
                break;
            case 4:
                return 'Loved';
                break;
        }
    }
}

module.exports = BeatmapsUtils;
