// osuapi-js/utils/mods.js
const modsconsts = require('../constants/modsConsts');

class osuMod {
    constructor(mod) {
        this.full  = modsconsts.codes[mod];
        this.alias = modsconsts.aliases[this.full];
    }
}

class Mods {
    static parseMods(m) {
        const enabled = [];

        const codes = modsconsts.codes;
        const values = Object.keys(codes).map(a => Number(a));
        for(let i = values.length-1; i >= 0; i--) {
            if(m >= values[i]) {
                m -= values[i];
                enabled.push(new osuMod(values[i]));
            }
            if(m == 0)
                break;
        }

        return enabled;
    }

    static calcMods(m) {
        if(!Array.isArray(m))
            throw new TypeError(m + ' is not an Array!');
        m = m.map(mod => mod.toUpperCase());
        
        let mods = 0;
        m.map(mod => {
            const codes = modsconsts.codes;
            const aliases = modsconsts.aliases;
            let alias = Object.keys(aliases).find(a => aliases[a] == mod);
            if(!alias)
                alias = Object.keys(codes).find(a => codes[a] == mod);
            else
                alias = Object.keys(codes).find(a => codes[a] == alias);
            if(!alias)
                throw new TypeError('There\'s no such mod as ' + mod);

            mods += alias;
        });

        return mods;
    }
}

module.exports = Mods;
